# unsupervised ml

## kmeans clustering
- outliers must be handled before kmeans clustering
- data must be scaled to ensure proper clustering
- Issues
  - k or number of clusters should be given before clustering


## heighrichal clustering
- 

In [2]:
import pandas as pd

# Content-based Recommender system
- rating based
- popularity based
  - cold start for a new user can be recommended by popularity
- content based
- user similarity based

In [1]:
# content based filtering
# USER PROFILE CREATION

doc = "Superheroes are often as popular as celebrities, as influential as politicians, and sometimes even as revered as gods. But that's when they're using their powers for good. What happens when the heroes go rogue and start abusing their powers? When it's the powerless against the super powerful, the Boys head out on a heroic quest to expose the truth about the Seven and Vought, the multibillion-dollar conglomerate that manages the superheroes and covers up their dirty secrets. Based on the comic book series of the same name."

print(len(set(doc.split())))

# USER PROFILE CREATION
# create a set of unique words in the document
# create a one hot encoding of the document using the set of unique words
# ie i have a ds with one row for one movie and i have a list of keywords for that movie. (i can use generas and other features as well)
# i then use my ratings and multiply to the one hot encoding to get a weighted one hot encoding.
# sum the weighted one hot encoding across all movies to get a score for each keyword.
# this is the preference score for each keyword for that user.



66


- drawback of one hot encoding because of which it is not used for movie vector creation
  - if a movie with 3 characters A,B,C has a rating
  - another movie with 3 characters A,C,D has another rating
  - then it is likely that the rating difference is because of C and D and not because of A and B
  - this is not captured by One hot encoding
  - <weighted one hot encoding is used for user profile creation> as discussed in the previous step


- Tf-Idf is used for movie vector creation (and not user profile)
  - written in notes


- i can compare using cosine similarity (angle bw two vectors) between user profile and any random movie vector for similarity

In [ ]:
# user similarity based recommendation

In [ ]:
# create a user item matrix with ratings of movies by users
ratings = pd.DataFrame({
    "harry potter": [5,5,1,4],
    "titanic":[1,2,5,2],
    "InterruptedError":[4,5,1,4],
    "narnia":[None,4,2,5]
}) 

index = ["user1", "user2", "user3", "user4"]
ratings.index = index
ratings

,harry potter,titanic,InterruptedError,narnia
user1,5,1,4,NaN
user2,5,2,5,4.0
user3,1,5,1,2.0
user4,4,2,4,5.0


In [ ]:
# Mean center the ratings by subtracting the mean rating for each user from their ratings
ratings_centered = ratings.sub(ratings.mean(axis=1), axis=0)
ratings_centered

,harry potter,titanic,InterruptedError,narnia
user1,1.666667,-2.333333,0.666667,NaN
user2,1.000000,-2.000000,1.000000,0.00
user3,-1.250000,2.750000,-1.250000,-0.25
user4,0.250000,-1.750000,0.250000,1.25


In [19]:
cent = ratings.apply(lambda x: x - x.mean(), axis=1)

In [20]:
cent

,harry potter,titanic,InterruptedError,narnia
user1,1.666667,-2.333333,0.666667,NaN
user2,1.000000,-2.000000,1.000000,0.00
user3,-1.250000,2.750000,-1.250000,-0.25
user4,0.250000,-1.750000,0.250000,1.25


In [18]:
ratings.apply(lambda x: print(x.mean()))


3.75
2.5
3.5
3.6666666666666665


harry potter        None
titanic             None
InterruptedError    None
narnia              None
dtype: object

In [ ]:
cent # this is the mean centered ratings. we can then use this to calculate similarity between users and make recommendations based on that.

,harry potter,titanic,InterruptedError,narnia
user1,1.666667,-2.333333,0.666667,NaN
user2,1.000000,-2.000000,1.000000,0.00
user3,-1.250000,2.750000,-1.250000,-0.25
user4,0.250000,-1.750000,0.250000,1.25


In [25]:

# i dont want narnia to make predection
pred = cent.loc[:,["harry potter", "titanic", "InterruptedError"]]
pred

,harry potter,titanic,InterruptedError
user1,1.666667,-2.333333,0.666667
user2,1.000000,-2.000000,1.000000
user3,-1.250000,2.750000,-1.250000
user4,0.250000,-1.750000,0.250000


In [ ]:
# calculating cosine similarity between users based on the mean centered ratings and making recommendations based on that.
similarity = pred.dot(pred.T)
similarity

,user1,user2,user3,user4
user1,8.666667,7.0,-9.333333,4.666667
user2,7.000000,6.0,-8.000000,4.000000
user3,-9.333333,-8.0,10.687500,-5.437500
user4,4.666667,4.0,-5.437500,3.187500


In [32]:
import numpy as np

In [ ]:
from sklearn.metrics.pairwise import cosine_similarity


In [ ]:
sim = cosine_similarity(pred)



In [37]:
sim

array([[ 1.        ,  0.97072534, -0.96977876,  0.88788279],
       [ 0.97072534,  1.        , -0.99902487,  0.91465912],
       [-0.96977876, -0.99902487,  1.        , -0.93161422],
       [ 0.88788279,  0.91465912, -0.93161422,  1.        ]])

In [33]:
# Step 1: Your starting point (raw dot product)
# (Assuming pred is already filled with zeros via .fillna(0))
similarity = pred.dot(pred.T)

# Step 2: Calculate the magnitude (norm) for each user vector
# This gives you a single value per user representing their "vector length"
magnitudes = np.linalg.norm(pred, axis=1)

# Step 3: Normalize the similarity matrix
# np.outer(magnitudes, magnitudes) creates a grid where position (i, j) 
# is equal to (magnitude of user_i * magnitude of user_j)
cosine_sim_matrix = similarity / np.outer(magnitudes, magnitudes)

In [ ]:
cosine_sim_matrix # this is basically cos theeta in the dot product used for vectors

,user1,user2,user3,user4
user1,1.000000,0.970725,-0.969779,0.887883
user2,0.970725,1.000000,-0.999025,0.914659
user3,-0.969779,-0.999025,1.000000,-0.931614
user4,0.887883,0.914659,-0.931614,1.000000


In [43]:
sim = cosine_similarity(pred)
sim = pd.DataFrame(sim, index=pred.index, columns=pred.index)
k2 = sim["user1"].sort_values(ascending=False)
k2

user1    1.000000
user2    0.970725
user4    0.887883
user3   -0.969779
Name: user1, dtype: float64

In [47]:
sum_weighted_ratings = (k2.loc["user2"] * cent.loc["user2", "narnia"]) + (k2.loc["user4"] * cent.loc["user4", "narnia"])

In [49]:
# this is the predicted rating for user1 for narnia based on the ratings of user2 and user4 and their similarity to user1.
sum_weighted_ratings/(k2.loc["user2"] + k2.loc["user4"])

np.float64(0.5971422738143606)

In [50]:
4/3

1.3333333333333333

# Association Mining
1. Before calculating the rules, you need the Support values (how often things appear in the total of 6 transactions)
  - Milk: Appears in 4 out of 6 transactions (66.7%).
  - Butter: Appears in 4 out of 6 transactions (66.7%).
  - {Milk, Butter}: Both appear together in 3 out of 6 transactions (50%).
2. Confidence: when x what howmany times y was bought
   1. Confidence measures the strength of the implication. 
   2. It asks: Of the people who bought Milk, what percentage also bought Butter?
   3. Formula:() Support{Milk intersection Butter})/(Support{Milk})
   4. Calculation: 0.5/0.667 = 0.75 => (75%)
   5. Meaning: If a customer has Milk in their cart, there is a 75% chance they will also buy Butter.
3. Lift: Lift tells you if the relationship is actually meaningful or just a coincidence based on how popular Butter is anyway
   1. Formula: $\frac{\text{Confidence}(\text{Milk} \rightarrow \text{Butter})}{\text{Support}(\text{Butter})
   2. Calculation: $\frac{0.75}{0.667} = 1.125
   3. Meaning: Since the Lift is greater than 1, there is a positive association. Buying Milk makes a customer 1.125x more likely to buy Butter than an average customer chosen at random.
4. Key Takeaways:Confidence can be misleading if the second item (Butter) is very popular on its own.Lift is the "sanity check"—if Lift was 1, it would mean Milk and Butter are bought independently, and the "75% confidence" was just because everyone buys Butter regardless.
5. Leverage
6. Conviction: Strength of 
   1. Conviction(X->Y) = (1-Support(Y))/(1-Confidence(X->Y))
   2. I dont want Y to be purchased regardless of X in bulk and hence the numerator is no of times Y was not bpought
   3. we want Y to be bought whenever X was purchased and hence we want the dinominator to be low where we have percentage of times x was bought and y was not bought

# Marked product associations


In [79]:
df = pd.read_csv("./session5/Market.csv")


In [80]:
df.head()

,"MILK,BREAD,BISCUIT"
0,"BREAD,MILK,BISCUIT,CORNFLAKES"
1,"BREAD,TEA,BOURNVITA"
2,"JAM,MAGGI,BREAD,MILK"
3,"MAGGI,TEA,BISCUIT"
4,"BREAD,TEA,BOURNVITA"


In [81]:
df = df.rename(columns={"MILK,BREAD,BISCUIT":"cart"})
df.head()


,cart
0,"BREAD,MILK,BISCUIT,CORNFLAKES"
1,"BREAD,TEA,BOURNVITA"
2,"JAM,MAGGI,BREAD,MILK"
3,"MAGGI,TEA,BISCUIT"
4,"BREAD,TEA,BOURNVITA"


In [82]:
df = df['cart'].str.split(",", expand=True)
df.head()

,0,1,2,3
0,BREAD,MILK,BISCUIT,CORNFLAKES
1,BREAD,TEA,BOURNVITA,None
2,JAM,MAGGI,BREAD,MILK
3,MAGGI,TEA,BISCUIT,None
4,BREAD,TEA,BOURNVITA,None


In [ ]:
list_items = 